### DISCUSSION TITLE
Abilities to run unmodified GNU/Linux binaries under Windows

### DISCUSSION THREAD

**Discussion by fishiatee** [danoon2/Boxedwine#36] [Feature Requests]
Continue discussion from [#32](https://github.com/danoon2/Boxedwine/issues/32).

**Comment by danoon2** [comment]
I created a demo of running XOrg with Boxedwine and a batch file that shows how to install a debian/buster/i386 package.

http://boxedwine.org/v/BoxedWineWithXOrg-v21-Win32.zip

This is currently a 32-bit Windows build of Boxedwine and thus is pretty slow with a laggy mouse.  But as a proof of concept, overall it seems to work pretty well.  The much faster multi-threaded 64-bit Boxedwine currently hangs around the time XOrg detects the mouse.  I wasn't able to quickly figure out why so I will have to come back and look at that some other time.

**Reply by fishiatee** [reply]
Thanks for that. I will quickly review it and give you feedback.

**Reply by fishiatee** [reply]
Tested, better than WSL1 a little bit. I can try to attempt doing some wonky stuff :)

**Comment by fishiatee** [comment]
![image](https://user-images.githubusercontent.com/74685931/142531624-c9b2ee07-2ae0-45ab-b2a7-a6915d4b4d89.png)
Neofetch works, but not perfect.

**Comment by danoon2** [comment]
Yeah, I've always wished Boxedwine had a console, but currently my main goal is getting old Windows games to work.  Maybe eventually that can include only Linux games/apps too which would then raise the priority of proper console support in the kernel emulation.

**Comment by fishiatee** [comment]
Yeah, but GUI apps is working, and that increased BoxedWine functionally a lot.

**Comment by fishiatee** [comment]
Although only a small percentage is working.

**Comment by fishiatee** [comment]
I have 2 questions.

1. Can I run custom rootfs (like a Arch one instead of Debian)?
2. Is the kernel in BoxedWine is a emulated or a fake kernel like WSL1? (And if emulated, how? Virtual Machine, or natively?)

**Reply by danoon2** [reply]
There is nothing preventing other distributions from running, but setting up a file system from scratch is a pain.  You can't just use a premade ISO.  My first file systems were made by hand where I downloaded the debian packages and unzipped them.  The worst part is how I implemented link files.  All the links in the packages, like library versions, had to be recreated.  My link format is a text file with the extention ".link"  My Java project now handles this for me now: see DebianPackage.java unpack method.

https://github.com/danoon2/Boxedwine/blob/master/tools/debianFileSystem/src/org/boxedwine/DebianPackage.java

As for the kernel, it is my own custom/fake kernel.  I implement the syscalls and other things needed like /dev and /proc

https://github.com/danoon2/Boxedwine/tree/master/source/kernel

When I upgraded from Debian 9 to Debian 10, I had to add more syscalls since Debian 10/libc made use of newer syscall, like openat instead of open.  I would expect the same thing if you tried any newer distro.

**Reply by fishiatee** [reply]
That's mean the kernel is somewhat like WSL1? 
If then, that's mean the Linux is running natively on Windows?
I found this [branch](https://github.com/danoon2/Boxedwine/tree/james/XOrg_64bit_work). Is this the branch that contain the XOrg build you just sent me?

**Comment by danoon2** [comment]
I'm not familiar with WSL1.

Linux is not running natively because of my CPU emulation.  If I could get my x64 build to work with XOrg, that would be about 25-50% of native speed.  The 32-bit is much much slower.  The 64-bit Windows version of Boxedwine is a multi-threaded binary translator, so it converts x86 code to x64 and adds a memory offset to all the memory instructions.  Boxedwine emulates the Linux processes all inside a single host process, so each emulated process has its own memory offset to a 4GB chunk of reserved memory.

I just merged my changes I needed to give you the 32-bit build into the master branch.  If you build the Win32 project with visual studio and add the define BOXEDWINE_EXPERIMENTAL_FRAME_BUFFER to the c/c++ preprocessor options, you should get the same build I gave you.

The branch you found is my work on trying to get x64 to run with XOrg.  So far it hangs when detecting the mouse.

**Reply by fishiatee** [reply]
Why you need to emulating a CPU? I think you can just emulate the Linux userspace?

**Reply by fishiatee** [reply]
That will be much more faster.

**Reply by danoon2** [reply]
Wine requires access to a pretty strict memory range, because win32 apps load at a specific memory address, the default is 0x00400000.  In Linux they do this with a custom preloader to guarantee the address range is free and not used by the system.  If I didn't use my custom memory offsets, I would have to implement a preload for every platform. On top of this, Wine starts multiple processes, like wineserver, and they all communicate using unix sockets.  I could emulate the sockets with pipes on Windows.  But that would be platform specific.  I like that Boxedwine can run in a single process, it makes things like putting in the browser possible.  To get native speed, perhaps some day I could remove the memory offset in favor of nested pages tables (NTP on AMD and EPT on Intel) for the x86 platforms.  That is what VMWare does to get near native speed.

But since Wine already runs on x86 Linux and Mac, and Boxedwine isn't that useful on Windows except for the older games/apps the current version of Windows can't run.  My main focus this last year it getting Boxedwine to run fast on ARM64.  Currently I have a x86 to ARM64 binary translator that is pretty fast, faster than QEMU.  But it still needs work.  MDK performance test reports on the Mac Mini M1, my ARM code is about 4x faster than a Pentium 200.  So its far from running modern code, but a nice start.  See

https://www.vogons.org/viewtopic.php?p=970555#p970555

**Comment by fishiatee** [comment]
Okay I was poking with this a while and I guess you will need to use atleast Debian Bullseye or newer. Since Buster is pretty old

**Comment by Klubuntu** [comment]
HI, how is your progress with creating boxedwine only with the terminal version, it would be very useful for me also in the x64 version

**Comment by danoon2** [comment]
I will probably try to get Boxedwine x64 version to run XOrg since I'm curious why it doesn't.  But as for a proper terminal, I have no plans to work on that in the near feature.    My goal for the last year has been to get a Mac, Windows and Linux build that works for most games from the year 2000 and before.  So far I'm pretty close to this goal, I'm just have a few issues with OpenGL on Mac which is why I'm experimenting with Vulkan right now.  After that I plan to finish my Arm64 work and create a Mac M1 native build.

**Reply by Klubuntu** [reply]
How can I prepare my own version of BoxedWine with my Debian x86 based rootfs

**Reply by danoon2** [reply]
@Klubuntu   If your goal is to run a Linux app using Boxedwine, you will probably run into bugs.  If your Debian version is newer than 10 then I would expect there to be newer syscalls that I did not implement yet.  It is all possible to do, it would just require a good amount of development time from you.

If you just want to add a few packages to what I created, that should be easier.  You can start with

http://boxedwine.org/v/BoxedWineWithXOrg-v21-Win32.zip

In there is an example of adding a package.

After you added the package, you can just copy the files written in the root fs back into the debian zip file system if you want.

**Reply by Klubuntu** [reply]
@danoon2 how can I add packages without apt and the terminal doesn't work for me

**Reply by Klubuntu** [reply]
I'm problem if add rootfs to this zip, no work command using BoxedWine_console.exe

```
C:\Users\Klubuntu\Downloads\Compressed\BoxedWineWithXOrg-v21-Win32>boxedwine\Boxedwine_console -root fs -zip boxedwine\Debian10XOrg2-rootfs2.zip /bin/echo test
Starting ...
Command line arguments: "boxedwine\Boxedwine_console" "-root" "fs" "-zip" "boxedwine\Debian10XOrg2-rootfs2.zip" "/bin/echo" "test"
Using root directory: fs
Using zip file system: boxedwine\Debian10XOrg2-rootfs2.zip

[process exited with code 3221226505 (0xc0000409)]

```

**Reply by Klubuntu** [reply]
Sorry i have problem using chroot

```
C:\Users\Klubuntu\Downloads\Compressed\BoxedWineWithXOrg-v21-Win32>boxedwine\Boxedwine_console -root fs -zip boxedwine\Debian10XOrg2.zip  /usr/sbin/chroot /debian-rootfs /bin/bash
Starting ...
Command line arguments: "boxedwine\Boxedwine_console" "-root" "fs" "-zip" "boxedwine\Debian10XOrg2.zip" "/usr/sbin/chroot" "/debian-rootfs" "/bin/bash"
Using root directory: fs
Using zip file system: boxedwine\Debian10XOrg2.zip
Loaded boxedwine\Debian10XOrg2.zip in 618 ms
Launching "/usr/sbin/chroot" "/debian-rootfs" "/bin/bash"
/usr/sbin/chroot: cannot change root directory to '/debian-rootfs': Function not implemented
Boxedwine has shutdown
```

**Reply by danoon2** [reply]
@Klubuntu apt does not work, currently only dpkg works and its a pain to use because you have to install the dependencies yourself, it won't handle it.  Also you will need to manually download the packages and places them in the file system.

I'm not surprised to hear that chroot doesn't work.   I wonder if that relies on a syscall I did not implement.    If you looks at my syscall list, you can see that many entries are missing.  Also, you should use the command line argument: -uid 0   This will make the kernel report that you are running as root.

https://github.com/danoon2/Boxedwine/blob/e2021c45923555429fae19a79e5dc5e9761cc0f8/source/kernel/syscall.cpp#L1554

**Comment by fishiatee** [comment]
@danoon2 How to impletement a syscall or a function (the easy way)?

**Comment by fishiatee** [comment]
@Klubuntu There are many old abandoned projects that has the same goal, most recent one (but now seems dead) is Foreign Linux. Consider checking it out.
https://github.com/wishstudio/flinux

**Reply by fishiatee** [reply]
There also LINE, LINA, LBW, KEOW, coLinux, andLinux and UMLWin32 too. They are all dead projects, but may be a good way to improve BoxedWine for Linux support.

**Reply by danoon2** [reply]
@ClickNinYT   Thanks for that list, I took a look at all of them.  Looks like none of them were that advanced, almost like a proof of concept, except coLinux.  coLinux looks like it could do a lot.  But I imagine coLinux to me is how most people feel about Boxedwine.  When I look at the coLinux code I don't even know where to start :)

**Reply by fishiatee** [reply]
coLinux use a kernel driver in order to run. But that kernel has not been ported to x86_64 for a long time. If you can do a similar thing for x86_64 you could benefits from many of cool stuff.

**Reply by fishiatee** [reply]
There also a "Dashboard" for porting coLinux to x86_64 too.
https://colinux.fandom.com/wiki/Dashboard_for_developing_a_64_bit_coLinux

**Reply by fishiatee** [reply]
You could also take the terminal interactive console code from Foreign Linux for example.

**Reply by fishiatee** [reply]
Maybe this will help: https://github.com/wishstudio/flinux/blob/master/src/common/termios.h

**Comment by danoon2** [comment]
foreign linux looks pretty interesting.  Their console implementation is specific to running on Windows and I'd rather create some platform independent code, but it might be a good place to start.

@ClickNinYT  as for implementing a missing syscall in Boxedwine, I would start by looking at my syscall table

https://github.com/danoon2/Boxedwine/blob/e2021c45923555429fae19a79e5dc5e9761cc0f8/source/kernel/syscall.cpp#L1554

The file system code can get pretty complicated because the file handles belong to the process and those point to the kernel file objects which can be shared across processes (KFile.cpp), and the actual file I/O is in the node classes since the node could be a virtual file, file on the host system or in a zip file.

If you want to discuss a particular syscall, I might be able to give some hints about how to implement it.

**Comment by danoon2** [comment]
@ClickNinYT  I created an issue to track x64 not working for the demo I created in the begging on this discussion. https://github.com/danoon2/Boxedwine/issues/41

**Reply by fishiatee** [reply]
Can you try some other DE? Such as KDE (probably older version like KDE 3.5) or GNOME. Also where I can look for syscall? Like I want to impletement what's missing in order to run Debian 11 Bullseye. Since Debian 10 is pretty old in my opinion.

**Comment by fishiatee** [comment]
I also discovered about "atratus" project. Again it's a project that's has the same goal.
http://atratus.org/

**Reply by fishiatee** [reply]
I think 64Bit XOrg is working now (the XOrg_64bit_work branch), so now when I can except atleast an experimental console support?

**Reply by danoon2** [reply]
Thanks for testing.  I'm still trying to figure out why the mouse looses movement and clicks sometimes.  I tried using Debian 11 to see if that fixed it, but it didn't.

I'm not sure what my plans are for the console.  I might research what it would take to add pty support to my kernel emulator so that XOrg could run a terminal.  But if I think it would take more than a day or two I would probably just lower the priority of the console feature.  There are just too many things I want to do with this project and not enough time.

**Comment by fishiatee** [comment]
By the way when I compiling BoxedWine from Visual Studio 2022 (and of course downloading pre built binaries from your jenkins server too), there was only BoxedWine.exe, and no BoxedWineConsole.exe. How I can attempt to launch Linux stuff?

**Comment by danoon2** [comment]
In the Visual Studio project, there is an option in "Project (Top Menu)/ Properties (Sub Menu) / Linker / System / SubSystem", you can change this from windows to console.  You can also use the command line "editbin.exe /subsystem:console Boxedwine.exe" to change it to use console.  editbin is in the build tools somewhere in your visual studio folder.

I will now include both Boxedwine.exe and Boxedwine_console.exe in the auto build.zip files.

**Comment by fishiatee** [comment]
Will implementing Async IO syscall (syscall 245 - 249) help anything?

**Reply by fishiatee** [reply]
https://man7.org/linux/man-pages/man2/io_setup.2.html
https://man7.org/linux/man-pages/man2/io_submit.2.html
https://man7.org/linux/man-pages/man2/io_cancel.2.html
https://man7.org/linux/man-pages/man2/io_getevents.2.html
https://man7.org/linux/man-pages/man2/io_destroy.2.html

**Comment by danoon2** [comment]
I'm not aware of these syscalls blocking any programs from running.  As far as performance improvements if they were implemented, I guess that would depend on the program.  Currently I have no plans for these.  I'm pretty much just moving from game to game to see what I need to do to make it work and if this comes up then it will suddenly be a high priority.

**Comment by fishiatee** [comment]
I'm adding some syscalls and trying to adding console support, improve the emulated kernel at my fork. 
Currently I'm just really starting it and added some syscalls into the syscall table here:
https://github.com/ClickNinYT/Boxedwine/commit/d105c02f68786cd90eb51bc592035503bf5a4977

**Reply by fishiatee** [reply]
I'm curious what is the replacement of this code?
```
return _ENOSYS;
```

**Reply by fishiatee** [reply]
I'm also planning to implement `fork()` syscall, from my knowledge it was used to create process or something.

**Reply by fishiatee** [reply]
Also, `ptrace()` syscall is not properly implemented I guess, which is used for something like User Mode Linux, proot or chroot.

**Reply by danoon2** [reply]
All the error codes should already be mapped, I just put a K in front of them, so if you want to return _ENOSYS, then you would return K_ENOSYS.  Keep in mind that most syscalls return the negative of the error code to indicate to libc there was an error.  So to return _ENOSYS from a syscall, you would need to return -K_ENOSYS.  For example I return this for inotify syscalls

https://github.com/danoon2/Boxedwine/blob/60bec158ad8860376790a07375d4625a556fd778/source/kernel/syscall.cpp#L1259

As for libc's fork() call, libc will call clone syscall to implement fork and clone is implemented well enough for fork()
https://github.com/danoon2/Boxedwine/blob/60bec158ad8860376790a07375d4625a556fd778/source/kernel/kprocess.cpp#L1614

**Comment by fishiatee** [comment]
Okay unrelated to this topic but I think BoxedWine can be useful to run older 32bit software (such as coLinux, Atracus, LINE, Services for UNIX, etc), does it possible. Does Wine support Kernel-mode driver (that coLinux uses)?

**Reply by danoon2** [reply]
I'm not that familiar with coLinux, but I think their driver was needed mainly to set FS and GS registers on the CPU.  Since Boxedwine emulates those, the driver isn't necessary, though coLinux might still need it for normal functioning.  I'm not aware of Wine supporting drivers like this.

I'm not sure how getting a real Linux kernel into Boxedwine might work.  I assume the Linux kernel handles thread scheduling, setting up page tables, etc and this is all stuff that isn't compatible with Boxedwine.  I wonder of coLinux modified those parts of the Linux kernel to defer to Windows.

**Comment by fishiatee** [comment]
Any progress on Console Emulation yet? 
(well, VT100 console emulation might be good, right?)

**Comment by danoon2** [comment]
My understanding is that I would have to implement pty emulation in my fake kernel

https://man7.org/linux/man-pages/man7/pty.7.html

Then the terminal apps in x11 should work.  To get a terminal to work without x11, a UI would have be written for the terminal.  I see lots of examples on github of creating the terminal/slave side, but not of the kernal/master side (pty).

I'm not sure how much work it would be.  But currently I'm in the middle of 2 other features I would need to finish first.  ARMv8 support and Vulkan pass through.  I'm guessing those will take me quite a while to finish.

**Comment by fishiatee** [comment]
I had an idea, using the concept of User Mode Linux (UML) to make a real Linux kernel run inside BoxedWine's fake kernel. But the only requirement is that the kernel will need to have `ptrace()` syscall fully implemented.

And, how I can compile and adding more packages in your Java Root File System generator? I would like to add in Meson, Ninja, Python, Git, GCC and some libraries to compile and experimenting with that concept.

**Reply by fishiatee** [reply]
That will help you eliminating compatibility problems and reduce workloads.

**Reply by danoon2** [reply]
Most information on User Mode Linux is from the perspective of using it.  I'm not sure what it would take to make my fake kernel use it.  I see there are patches that allow it in the Linux Kernel and they aren't that big, which is good news.  I'm also not sure how UML really works, for example what does the UML kernel do with its own file system syscalls.  It appears it can use a block device and manage its own file system (I don't want file system image files) or can use the host file system (hostfs) as a virtual file system which I assume will forward the syscalls?  For debugging a new file system, I can see how UML would be very useful, but I don't think it will help out Boxedwine that much if it is just forwarding syscalls.

I also looked at ptrace, https://stackoverflow.com/questions/23928530/how-does-ptrace-work-in-linux, being able to hook into the syscalls shouldn't be that hard, but I imagine it would take a while to work out all the exact details.  

I added some documentation for the debian file system creator, let me know if something isn't clear

https://github.com/danoon2/Boxedwine/blob/master/tools/debianFileSystem/howto.txt

**Reply by danoon2** [reply]
One more thing, I see that you mentioned git.  Boxedwine has network support, but I have noticed that it doesn't always work.  Its on my list of things to look at after Vulkan and ARMv8.

**Reply by fishiatee** [reply]
It should help Boxedwine eliminating compatibility problems with applications. Because User Mode Linux is actually included in official Linux kernel source tree after release 2.6, compiling it should be pretty forward, just `make ARCH=um defconfig` and then `make ARCH=um` and you should be ready to go.
When you get UML running, the compatibility problems will be resolved. And thus so reduce workloads and the amount of works needed to get a Linux program running. The syscalls will be never missing anymore. And, if you can't get ptrace(), there must be a way to remove ptrace() requirement, mainly by removing all of it's reference, and then reimplement it in the source tree.

**Reply by danoon2** [reply]
I agree, if I could get UML to work, that would definitely be the way to go.  I will add this to the list of things to research in the future.  For now, I'm just focused on Vulkan and ARMv8.  Currently as part Vulkan, I've been trying to get Wine 7 to work.  Wine 7 requires the use of Linux shared memory.   So if I can get shared memory to work that should help the overall Linux functionality as well.

**Comment by Klubuntu** [comment]
**Can You help me**: Discussion https://github.com/danoon2/Boxedwine/discussions/50
